File: src/model/ModelValidator.js
const StepTypes = require('./steps/StepTypes')
const ErrorMessages = require('../constants/ErrorMessages')

class ModelValidator {
    constructor() {
        this._rules = []
        this.initRules()
    }

    everyProcessStepMustHaveSourceAndTarget(model) {
        const result = []
        const processSteps = this.getProcessSteps(model)

        processSteps.forEach(step => {
            if (step.sourceConnections.length === 0) {
                result.push(this.formatErrorMessage(ErrorMessages.NO_SOURCE_CONNECTION, step))
            }
            if (step.targetConnections.length === 0) {
                result.push(this.formatErrorMessage(ErrorMessages.NO_TARGET_CONNECTION, step))
            }
        })

        return result
    }

    formatErrorMessage(message, step) {
        return `\r\n${message}\r\n編號: ${step.id}\r\n程式碼: ${step.code}`
    }

    getProcessSteps(model) {
        return model.steps.filter(
            step => step.type !== StepTypes.START && step.type !== StepTypes.STOP
        )
    }

    initRules() {
        this._rules.push(this.mustHaveStartAndStop.bind(this))
        this._rules.push(this.everyProcessStepMustHaveSourceAndTarget.bind(this))
    }

    mustHaveStartAndStop(model) {
        const result = []
        if (!model.findEntryPoint()) {
            result.push(ErrorMessages.MISSING_START_STEP)
        }
        if (!model.findExitPoint()) {
            result.push(ErrorMessages.MISSING_STOP_STEP)
        }
        return result
    }

    validate(model) {
        if (!model) {
            throw new Error('模型不能為空')
        }

        let messages = []
        this._rules.forEach(rule => {
            messages = messages.concat(rule(model))
        })
        return messages
    }
}

module.exports = ModelValidator

File: Models/ModelValidator.cs
using WorkflowEngine.Models.Steps;

namespace WorkflowEngine.Models;

public class ModelValidator
{
    public List<string> Validate(Model model)
    {
        var errors = new List<string>();

        // 檢查是否有開始節點
        if (!model.Steps.Any(s => s.Type == StepTypes.START))
        {
            errors.Add("找不到開始節點");
        }

        // 檢查是否有結束節點
        if (!model.Steps.Any(s => s.Type == StepTypes.STOP))
        {
            errors.Add("找不到結束節點");
        }

        // 檢查是否有孤立的節點
        foreach (var step in model.Steps)
        {
            if (step.Type != StepTypes.START && !model.Connections.Any(c => c.Target == step))
            {
                errors.Add($"節點 {step.Label} 沒有輸入連線");
            }

            if (step.Type != StepTypes.STOP && !model.Connections.Any(c => c.Source == step))
            {
                errors.Add($"節點 {step.Label} 沒有輸出連線");
            }
        }

        return errors;
    }
}